test/test_array.cpp

```diff
@@ -1559,4 +1560,525 @@ NONCONCURRENT_TEST(Array_count)
+    }
+
+    // now simulate the compression of a in less bits
+    uint64_t buff[2] = {0, 0};
+    BfIterator it(buff, 0, 5, 5, 0); // 5 bits because 4 bits for the values + 1 bit for the sign
+    for (size_t i = 0; i < 16; ++i) {
+        it.set_value(i);
+        CHECK_EQUAL(*it, a[i]);
+        ++it;
+    }
+    struct MyClass {
+        uint64_t* _data;
+        int64_t get(size_t ndx) const
+        {
+            BfIterator it(_data, 0, 5, 5, ndx);
+            return sign_extend_value(5, *it);
+        }
+    };
+    // a bit of set up here.
+    MyClass my_class;
+    my_class._data = buff;
+    using Fetcher = impl::CompressedDataFetcher<MyClass>;
+    Fetcher my_fetcher;
+    my_fetcher.ptr = &my_class;
+
+    // verify that the fetcher returns the same values
+    for (size_t i = 0; i < size; ++i) {
+        CHECK_EQUAL(my_fetcher.ptr->get(i), a[i]);
+    }
+
+    std::vector<size_t> diffent_width_lb;
+    for (const auto v : a) {
+        const auto pos = lower_bound((const char*)buff, size, v, my_fetcher);
+        diffent_width_lb.push_back(pos);
+    }
+
+    CHECK_EQUAL(expected_default_lb, diffent_width_lb);
+}
+
+TEST(UpperBoundCorrectness)
+{
+    constexpr auto size = 16;
+    int64_t a[size]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15};
+    std::vector<size_t> expected_default_ub;
+    for (const auto v : a) {
+        const auto pos = upper_bound<64>((const char*)(&a), size, v);
+        expected_default_ub.push_back(pos);
+    }
+
+    // now simulate the compression of a in less bits
+    uint64_t buff[2] = {0, 0};
+    BfIterator it(buff, 0, 5, 5, 0); // 5 bits because 4 bits for the values + 1 bit for the sign
+    for (size_t i = 0; i < size; ++i) {
+        it.set_value(i);
+        CHECK_EQUAL(*it, a[i]);
+        ++it;
+    }
+    struct MyClass {
+        uint64_t* _data;
+        int64_t get(size_t ndx) const
+        {
+            BfIterator it(_data, 0, 5, 5, ndx);
+            return sign_extend_value(5, *it);
+        }
+    };
+    // a bit of set up here.
+    MyClass my_class;
+    my_class._data = buff;
+    using Fetcher = impl::CompressedDataFetcher<MyClass>;
+    Fetcher my_fetcher;
+    my_fetcher.ptr = &my_class;
+
+    // verify that the fetcher returns the same values
+    for (size_t i = 0; i < size; ++i) {
+        CHECK_EQUAL(my_fetcher.ptr->get(i), a[i]);
+    }
+
+    std::vector<size_t> diffent_width_ub;
+    for (const auto v : a) {
+        const auto pos = upper_bound((const char*)buff, size, v, my_fetcher);
+        diffent_width_ub.push_back(pos);
+    }
+
+    CHECK_EQUAL(expected_default_ub, diffent_width_ub);
+}
+
+TEST(ParallelSearchEqualMatch)
+{
+    std::mt19937_64 gen64;
+    constexpr size_t buflen = 4;
+    uint64_t buff[buflen];
+    std::vector<int64_t> values;
+    for (size_t width = 1; width <= 64; width++) {
+        const size_t size = (buflen * 64) / width;
+        const uint64_t bit_mask = 0xFFFFFFFFFFFFFFFFULL >> (64 - width); // (1ULL << width) - 1;
```

Comment:
So that is supposed to be more readable? :-)

Reply:
This is how we do it in` ArrayWithFind::find_zero`, 
`uint64_t mask = (width == 64 ? ~0ULL : ((1ULL << (width == 64 ? 0 : width)) - 1ULL));`
even less readable :-D